File: python/src/scmarkeragent/uberon_ontology.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""UBERON anatomy-ontology helpers: tissue subsumption for the tissue_root switch.

Mirrors cl_ontology.py in style and normalization, but:
  - parses uberon-basic.obo (id: UBERON:..., is_a: UBERON:..., and
    relationship: part_of UBERON:...) so a *parent* edge = is_a OR part_of.
    part_of is what makes "colon" reachable from "intestine"; is_a covers
    anatomical subtype edges. Non-UBERON targets (CARO/BFO) are ignored.
  - adds descendants(id) / members(root) so a query tissue can pool every
    same-organ anatomical sub-region under a root (e.g. intestine -> colon,
    ileum, Peyer's patch, vermiform appendix ...).

Deterministic, file-only: no network, no LLM. Both arms parse the obo at
runtime, so the ontology file is the single source of truth (no committed
derived artifact to drift).
"""

import re
from .config import OBO_UBERON as OBO_DEFAULT

_DASH = re.compile(r"[\u2010-\u2015\u2212]")
_PAREN_TAIL = re.compile(r"\s*\([^)]*\)\s*$")
_WS = re.compile(r"\s+")
_SP_HYPHEN = re.compile(r" ?- ?")
_SYN = re.compile(r'^synonym: "([^"]*)" *([A-Z]*).*$')
_ISA = re.compile(r"^is_a:\s*(UBERON:\d+)")
_PARTOF = re.compile(r"^relationship:\s*part_of\s+(UBERON:\d+)")


def norm_name(x):
    if x is None:
        return ""
    x = str(x).strip().lower()
    x = _PAREN_TAIL.sub("", x)
    x = _DASH.sub("-", x)
    x = _WS.sub(" ", x)
    x = _SP_HYPHEN.sub("-", x)
    return x.strip()


class UberonOntology:
    def __init__(self, obo=OBO_DEFAULT):
        self.id2name = {}
        self.parents = {}  # id -> [is_a + part_of parents]
        self._children = {}  # id -> [children] (inverse of parents)
        self._anc = {}  # ancestor cache
        self._desc = {}  # descendant-set cache
        self.namemap = {}  # normalized name/EXACT-synonym -> id
        self._parse(obo)
        self._invert_children()
        self._build_namemap_from_parsed()

    # ---- parse ----------------------------------------------------------------
    def _parse(self, obo):
        prim: dict[str, set[str]] = {}
        syn: dict[str, set[str]] = {}
        cid = None
        name = None
        obs = False
        syns: list[str] = []

        def flush():
            if cid is None or obs:
                return
            if name is not None:
                prim.setdefault(norm_name(name), set()).add(cid)
            for s in syns:
                syn.setdefault(norm_name(s), set()).add(cid)

        with open(obo, encoding="utf-8", errors="replace") as fh:
            for raw in fh:
                x = raw.rstrip("\n")
                if x == "[Term]":
                    flush()
                    cid = None
                    name = None
                    obs = False
                    syns = []
                    continue
                if x.startswith("id: UBERON:"):
                    cid = x[4:].strip()
                elif x.startswith("name: "):
                    name = x[6:]
                    if cid is not None:
                        self.id2name[cid] = name
                elif x.startswith("is_obsolete: true"):
                    obs = True
                elif x.startswith("is_a:"):
                    m = _ISA.match(x)
                    if m and cid is not None:
                        self.parents.setdefault(cid, []).append(m.group(1))
                elif x.startswith("relationship: part_of"):
                    m = _PARTOF.match(x)
                    if m and cid is not None:
                        self.parents.setdefault(cid, []).append(m.group(1))
                elif x.startswith("synonym: "):
                    m = _SYN.match(x)
                    if m and m.group(1) and m.group(2) == "EXACT":
                        syns.append(m.group(1))
        flush()
        self._prim = prim
        self._syn = syn

    def _build_namemap_from_parsed(self):
        prim_v = {k: next(iter(v)) for k, v in self._prim.items() if len(v) == 1}
        syn_v = {
            k: next(iter(v))
            for k, v in self._syn.items()
            if len(v) == 1 and k not in prim_v
        }
        nm = dict(prim_v)
        nm.update(syn_v)
        self.namemap = nm

    def _invert_children(self):
        for k, ps in self.parents.items():
            for p in set(ps):
                self._children.setdefault(p, []).append(k)

    # ---- graph ops ------------------------------------------------------------
    def ancestors(self, uid):
        if not uid:
            return set()
        if uid in self._anc:
            return self._anc[uid]
        out = set()
        frontier = list(self.parents.get(uid, []))
        while frontier:
            nxt = []
            for f in frontier:
                if f not in out:
                    out.add(f)
                    nxt.extend(self.parents.get(f, []))
            frontier = nxt
        self._anc[uid] = out
        return out

    def descendants(self, uid):
        """All ids reachable downward via is_a/part_of children (excludes uid)."""
        if not uid:
            return set()
        if uid in self._desc:
            return self._desc[uid]
        seen = set()
        frontier = list(self._children.get(uid, []))
        while frontier:
            nxt = []
            for f in frontier:
                if f not in seen:
                    seen.add(f)
                    nxt.extend(self._children.get(f, []))
            frontier = nxt
        self._desc[uid] = seen
        return seen

    def is_desc(self, a, b):
        """a is a descendant of b (b is ancestor of a)."""
        return bool(a) and bool(b) and (b in self.ancestors(a))

    def closure_self(self, uid):
        """{uid} + all descendants -- the set to pool for a root tissue."""
        if not uid:
            return set()
        return {uid} | self.descendants(uid)

    # ---- name <-> id ----------------------------------------------------------
    def name_of(self, uid):
        return self.id2name.get(uid, uid)

    def resolve(self, name):
        """Normalized name / EXACT synonym -> UBERON id (None if unresolved).
        Accepts a raw 'UBERON:xxxxxxx' id directly."""
        if not name:
            return None
        s = str(name).strip()
        if s.upper().startswith("UBERON:"):
            return s.upper()
        return self.namemap.get(norm_name(s))

    def members(self, root):
        """Resolve a root tissue (name or id) and return {root_id} + descendant
        ids. Empty set if the root name cannot be resolved."""
        rid = self.resolve(root)
        if not rid:
            return set()
        return self.closure_self(rid)


if __name__ == "__main__":
    import time

    t0 = time.time()
    ub = UberonOntology()
    print(f"parsed uberon-basic.obo in {time.time() - t0:.1f}s")
    print(f"  terms (id2name): {len(ub.id2name)}")
    print(f"  parents entries: {len(ub.parents)}")
    print(f"  namemap entries: {len(ub.namemap)}")
    for nm in ["intestine", "brain", "kidney", "heart", "lung", "pancreas", "stomach"]:
        rid = ub.resolve(nm)
        mem = ub.members(nm)
        kids = sorted(ub.name_of(d) for d in ub.descendants(rid)) if rid else []
        print(
            f"  {nm!r:12s} -> {rid} ({ub.name_of(rid) if rid else 'NA'}); "
            f"members={len(mem)}; e.g. {kids[:6]}"
        )
    # subsumption sanity: intestine must NOT pull in stomach/esophagus
    it = ub.resolve("intestine")
    for probe in [
        "stomach",
        "esophagus",
        "colon",
        "ileum",
        "Peyer's patch",
        "vermiform appendix",
    ]:
        pid = ub.resolve(probe)
        print(f"  is_desc({probe!r:20s}, intestine) = {ub.is_desc(pid, it)}")

File: R/inst/scmarkeragent/rflow/uberon_ontology.R
#!/usr/bin/env Rscript
# =============================================================================
# uberon_ontology.R -- UBERON anatomy-ontology helpers for the R arm, with the semantics
# every arm shares so that all of them gate tissues identically.
#
# Parses uberon-basic.obo treating a *parent* edge as is_a OR relationship:
# part_of (part_of is what makes "colon" reachable from "intestine"; is_a covers
# anatomical subtype edges). Non-UBERON targets (CARO/BFO) are ignored. Exposes:
#   uberon_load(obo)            -> ontology object (list of maps + children)
#   uberon_resolve(ub, name)    -> UBERON:id (NA if unresolved; accepts raw id)
#   uberon_descendants(ub, id)  -> character() of downstream ids (excl. id)
#   uberon_closure_ids(ub, root)-> {root_id} + descendants (the pool set)
#   uberon_closure_names(ub, r) -> normalized names for that closure
#   uberon_name(ub, id)         -> term name
# Deterministic, file-only (no network/LLM); the obo is parsed at runtime.
# =============================================================================
suppressPackageStartupMessages({
  library(data.table)
})

# normalization IDENTICAL to cl_ontology.R .norm_name
.ub_norm <- function(x) {
  x <- tolower(trimws(as.character(x)))
  x <- gsub("\\s*\\([^)]*\\)\\s*$", "", x) # drop trailing "(...)"
  x <- gsub("[\u2010-\u2015\u2212]", "-", x) # unicode dashes -> hyphen
  x <- gsub("\\s+", " ", x) # collapse whitespace
  x <- gsub(" ?- ?", "-", x) # normalize spaced hyphens
  trimws(x)
}

uberon_term_blocks <- function(lines) {
  starts <- which(lines == "[Term]")
  if (!length(starts)) {
    return(list())
  }
  ends <- c(starts[-1] - 1L, length(lines))
  Map(function(first, last) lines[first:last], starts, ends)
}

parse_uberon_term <- function(block) {
  id_line <- block[startsWith(block, "id: UBERON:")]
  if (!length(id_line) || any(block == "is_obsolete: true")) {
    return(NULL)
  }
  ontology_id <- sub("^id: ", "", id_line[1])
  name_line <- block[startsWith(block, "name: ")]
  name <- if (length(name_line)) sub("^name: ", "", name_line[1]) else NA_character_
  edge_lines <- block[
    startsWith(block, "is_a:") |
      startsWith(block, "relationship: part_of")
  ]
  parents <- unique(unlist(regmatches(
    edge_lines, regexpr("UBERON:\\d+", edge_lines)
  )))
  synonym_lines <- block[startsWith(block, "synonym: ")]
  synonyms <- vapply(synonym_lines, function(line) {
    match <- regexec('^synonym: "([^"]*)" *(EXACT).*$', line)
    groups <- regmatches(line, match)[[1]]
    if (length(groups) >= 3L) groups[2] else NA_character_
  }, "")
  list(
    id = ontology_id,
    name = name,
    parents = parents[nzchar(parents)],
    synonyms = unique(synonyms[!is.na(synonyms) & nzchar(synonyms)])
  )
}

unambiguous_name_map <- function(terms) {
  primary <- list()
  synonyms <- list()
  for (term in terms) {
    if (!is.na(term$name)) {
      key <- .ub_norm(term$name)
      primary[[key]] <- c(primary[[key]], term$id)
    }
    for (synonym in term$synonyms) {
      key <- .ub_norm(synonym)
      synonyms[[key]] <- c(synonyms[[key]], term$id)
    }
  }
  unique_id <- function(values) {
    values <- unique(values)
    if (length(values) == 1L) values[1] else NA_character_
  }
  primary_values <- vapply(primary, unique_id, "")
  primary_values <- primary_values[!is.na(primary_values)]
  synonym_values <- vapply(synonyms, unique_id, "")
  synonym_values <- synonym_values[!is.na(synonym_values)]
  synonym_values <- synonym_values[
    !(names(synonym_values) %in% names(primary_values))
  ]
  c(primary_values, synonym_values)
}

invert_uberon_parents <- function(parents) {
  children <- list()
  for (child in names(parents)) {
    for (parent in unique(parents[[child]])) {
      children[[parent]] <- c(children[[parent]], child)
    }
  }
  lapply(children, unique)
}

uberon_load <- function(obo) {
  terms <- Filter(Negate(is.null), lapply(
    uberon_term_blocks(readLines(obo, warn = FALSE)),
    parse_uberon_term
  ))
  ids <- vapply(terms, function(term) term$id, "")
  names <- vapply(terms, function(term) term$name, "")
  parents <- setNames(lapply(terms, function(term) term$parents), ids)
  list(
    id2name = setNames(names, ids),
    parents = parents,
    children = invert_uberon_parents(parents),
    namemap = unambiguous_name_map(terms)
  )
}


uberon_name <- function(ub, id) {
  v <- if (!is.na(id)) ub$id2name[[id]] else NA
  if (is.null(v)) v <- NA
  if (is.na(v)) id else v
}

uberon_resolve <- function(ub, name) {
  if (is.null(name) || is.na(name) || !nzchar(name)) {
    return(NA_character_)
  }
  s <- trimws(as.character(name))
  if (grepl("^UBERON:", toupper(s))) {
    return(toupper(s))
  }
  hit <- ub$namemap[[.ub_norm(s)]]
  if (is.null(hit)) NA_character_ else hit
}

uberon_descendants <- function(ub, id) {
  if (is.null(id) || is.na(id)) {
    return(character(0))
  }
  seen <- character(0)
  frontier <- ub$children[[id]]
  if (is.null(frontier)) frontier <- character(0)
  while (length(frontier)) {
    nxt <- character(0)
    for (f in frontier) {
      if (!(f %in% seen)) {
        seen <- c(seen, f)
        children <- ub$children[[f]]
        if (!is.null(children)) nxt <- c(nxt, children)
      }
    }
    frontier <- unique(nxt)
  }
  unique(seen)
}

uberon_closure_ids <- function(ub, root) {
  rid <- uberon_resolve(ub, root)
  if (is.na(rid)) {
    return(character(0))
  }
  unique(c(rid, uberon_descendants(ub, rid)))
}

uberon_closure_names <- function(ub, root) {
  ids <- uberon_closure_ids(ub, root)
  if (!length(ids)) {
    return(character(0))
  }
  unique(.ub_norm(vapply(ids, function(i) uberon_name(ub, i), "")))
}

## The tissue set one context pools, as UBERON ids and normalized names. It lives here
## rather than in a stage script because three stages ask the same question -- candidate
## scoring, preprocessing's DE universe, and the source lookup -- and a copy per caller is
## how they stop agreeing about which tissues a run covers.
scma_tissue_closure <- function(ub, tissue, tissue_root) {
  tr <- tolower(trimws(if (!nzchar(tissue_root)) "self" else tissue_root))
  ids <- character(0)
  nms <- .ub_norm(tissue)
  rid <- uberon_resolve(ub, tissue)
  if (!is.na(rid)) ids <- rid
  if (tr %in% c("exact", "none", "off")) {
    return(list(ids = ids, names = nms))
  }
  root <- if (tr %in% c("self", "")) tissue else tissue_root
  cids <- uberon_closure_ids(ub, root)
  if (length(cids)) {
    ids <- unique(c(ids, cids))
    nms <- unique(c(nms, uberon_closure_names(ub, root)))
  }
  list(ids = ids, names = nms)
}
